### ISSUE TITLE
Gradient dropping is better compared to the baseline

### ISSUE THREAD

**Issue by afaji** [marian-nmt/marian-dev#96]
Not complaining, but now, if we use '--drop_rate 0.99', we'll get a better speed and convergence compared to without it.

Tested on 3 GTX 1080s

without:
```
[2017-07-30 21:32:36] Ep. 1 : Up. 1000 : Sen. 79760 : Cost 166.57 : Time 198.30s : 10025.85 words/s
[2017-07-30 21:35:50] Ep. 1 : Up. 2000 : Sen. 159760 : Cost 147.62 : Time 194.34s : 10287.65 words/s
[2017-07-30 21:39:07] Ep. 1 : Up. 3000 : Sen. 239760 : Cost 137.90 : Time 196.72s : 10180.63 words/s
...
```

with --drop_rate 0.99:
```
[2017-07-30 21:02:25] Ep. 1 : Up. 1000 : Sen. 79760 : Cost 163.44 : Time 199.07s : 9987.22 words/s
[2017-07-30 21:04:58] Ep. 1 : Up. 2000 : Sen. 159760 : Cost 135.93 : Time 152.16s : 13142.35 words/s
[2017-07-30 21:07:31] Ep. 1 : Up. 3000 : Sen. 239760 : Cost 123.69 : Time 153.47s : 13046.26 words/s
...
```

**Comment by afaji** [comment]
FYI @kpu

**Comment by emjotde** [comment]
Same seed? Faster update -> smaller staleness?

**Comment by kpu** [comment]
@dr-alan-turing @DianaDespa you saw something different?

**Comment by kpu** [comment]
Can you be more specific about model i.e. command to run on Valhalla?

**Comment by afaji** [comment]
same seed, same everything. 
The command:
´´´
../../build/marian-test \
 --model model/model.npz \
 --devices $GPUS --seed 321 \
 --train-sets data/corpus.bpe.ro data/corpus.bpe.en \
 --vocabs model/vocab.ro.yml model/vocab.en.yml \
 --dim-vocabs 66000 50000 \
 --mini-batch 80 \
# --drop-rate 0.99 \
 --layer-normalization --dropout-rnn 0.2 --dropout-src 0.1 --dropout-trg 0.1 \
 --early-stopping 50 \
 --valid-freq 10000 --save-freq 10000 --disp-freq 1000 \
 --valid-sets data/newsdev2016.bpe.ro data/newsdev2016.bpe.en \
 --valid-metrics cross-entropy valid-script \
 --valid-script-path ./scripts/validate.sh \
 --log model/train.log --valid-log model/valid.lo
´´´

This is from amunmt's run-me.sh (https://github.com/marian-nmt/marian/blob/marian-integration/examples/training/run-me.sh)
The weird thing is, not only gradient drop has better wps. it converges faster too.

**Comment by afaji** [comment]
This is my result frrom around 4 months ago:
Without drop:
```
#step loss
1000 153.75
2000 125.10
3000 108.12
4000 95.18
5000 86.32
6000 79.65
7000 74.17
8000 70.42
9000 67.42
10000 64.52
```

With droping 0.99
```
#step loss
1000 163.48
2000 135.40
3000 121.70
4000 109.85
5000 100.58
6000 92.76
7000 86.78
8000 82.10
9000 77.60
10000 74.53
```

(I don't have the time benchmark for that result, but it is clear that we get worse loss now)

**Comment by emjotde** [comment]
Can you try with a commit version from roughly that date and compare?

**Comment by DianaDespa** [comment]
I saw the number of words per second is bigger, but with quantization, the convergence is slowed down.

**Comment by emjotde** [comment]
Hi
Cannot confirm this. The progression I see with the above command is for `marian-dev` on zisa, 3 GPUs:

```
[2017-07-31 14:16:10] Ep. 1 : Up. 1000 : Sen. 79760 : Cost 152.56 : Time 188.15s : 10576.82 words/s
[2017-07-31 14:19:08] Ep. 1 : Up. 2000 : Sen. 159760 : Cost 123.33 : Time 178.83s : 11173.27 words/s
[2017-07-31 14:22:08] Ep. 1 : Up. 3000 : Sen. 239760 : Cost 107.14 : Time 179.83s : 11132.84 words/s
```

and for stable `marian` on the same machine:

```
[2017-07-31 14:46:31] Ep. 1 : Up. 1000 : Sen. 80000 : Cost 152.85 : Time 242.95s : 8216.08 words/s
[2017-07-31 14:49:50] Ep. 1 : Up. 2000 : Sen. 160000 : Cost 123.39 : Time 198.99s : 10050.63 words/s
[2017-07-31 14:53:11] Ep. 1 : Up. 3000 : Sen. 240000 : Cost 107.30 : Time 200.93s : 9967.96 words/s
```
Both with fresh pulls from respective `master`. Are you sure this isn't something you are getting on your own branch from pull request https://github.com/marian-nmt/marian-dev/pull/95 in which case this would be your own mess to fix.

**Comment by emjotde** [comment]
What's much more interesting is, why is `marian-dev` more than 10% faster than `marian`? :)

**Comment by afaji** [comment]
what I obtained is fresh from marian-dev's master (has nothing to do with my own PR)

did you get your result from master or from different branch?

**Comment by emjotde** [comment]
Both times from master as stated above.

**Comment by afaji** [comment]
issue found.
I got the correct result when I use the following on my .sh script :
```
../../build/marian \
 --model model/model.npz \
 --devices $GPUS --seed 321 \
 --train-sets data/corpus.bpe.ro data/corpus.bpe.en \
 --vocabs model/vocab.ro.yml model/vocab.en.yml \
 --dim-vocabs 66000 50000 \
 --mini-batch 80 \
 --layer-normalization --dropout-rnn 0.2 --dropout-src 0.1 --dropout-trg 0.1 \
 --early-stopping 50 \
 --valid-freq 10000 --save-freq 10000 --disp-freq 1000 \
 --valid-sets data/newsdev2016.bpe.ro data/newsdev2016.bpe.en \
 --valid-metrics cross-entropy valid-script \
 --valid-script-path ./scripts/validate.sh \
 --log model/train.log --valid-log model/valid.log
```

But not when doing this:
```
../../build/marian \
 --model model/model.npz \
 --devices $GPUS --seed 321 \
 --train-sets data/corpus.bpe.ro data/corpus.bpe.en \
 --vocabs model/vocab.ro.yml model/vocab.en.yml \
 --dim-vocabs 66000 50000 \
 --mini-batch 80 \
 # --drop-rate 0.99 \
 --layer-normalization --dropout-rnn 0.2 --dropout-src 0.1 --dropout-trg 0.1 \
 --early-stopping 50 \
 --valid-freq 10000 --save-freq 10000 --disp-freq 1000 \
 --valid-sets data/newsdev2016.bpe.ro data/newsdev2016.bpe.en \
 --valid-metrics cross-entropy valid-script \
 --valid-script-path ./scripts/validate.sh \
 --log model/train.log --valid-log model/valid.log
```

Stupid me. Sorry for the noise
